### PR TITLE
feat(ui): Provide in-game display of damage dropoff on weapons.

### PR DESCRIPTION
**Pull request by Quantumshark**
**Feature:**

## Feature Details
Displays the damage before and after the damage dropoff on weapons. Display is always `damage before dropoff - damage after dropoff`, rather than low-high, as this way it is visible whether the damage is increasing or decreasing with range.

Doesn't currently work with submunitions, and given that currently that is (as far as I am aware) the only place damage dropoff is currently used in vanilla, I'd really like to actually get that working.

## UI Screenshots
image
An energy blaster, with `"dropoff modifer" .5`

## Usage Examples
N/A

## Testing Done
Added `"dropoff modifer" .5` to the energy blaster, and checked that it was displaying correctly. Also checked that outfits without a damage dropoff weren't broken. Also noted that the flamethrower does not yet have any display of its dropoff.

### Automated Tests Added
N/A

## Performance Impact
Should be no noticeable performance impact, as this code only runs in menus and not in-flight.
No performance impact observed in menus.

### CODE REVIEW

File: source/OutfitInfoDisplay.cpp [hunk 1]
```diff
@@ -560,7 +565,13 @@ void OutfitInfoDisplay::UpdateAttributes(const Outfit &outfit)
...
 			{
 				attributeLabels.emplace_back(VALUE_NAMES[i].first + PER_SECOND);
-				attributeValues.emplace_back(Format::Number(60. * values[i] / reload) + VALUE_NAMES[i].second);
+				if(fullDropoff != 1 && VALUE_NAMES[i].first.find("damage") != string::npos)
+				{
+					attributeValues.emplace_back(Format::Number(60. * values[i] / reload) + VALUE_NAMES[i].second
+						+ " - " + Format::Number(60. * fullDropoff * values[i] / reload) + VALUE_NAMES[i].second);
+				}
+				else
+					attributeValues.emplace_back(Format::Number(60. * values[i] / reload) + VALUE_NAMES[i].second);
```
Comment (Hurleveur) [inline source/OutfitInfoDisplay.cpp:574]:
```suggestion
				attributeValues.emplace_back(Format::Number(60. * values[i] / reload) + VALUE_NAMES[i].second);
				if(fullDropoff != 1 && VALUE_NAMES[i].first.find("damage") != string::npos)
					attributesValues.back() += " - " + Format::Number(60. * fullDropoff * values[i] / reload) + VALUE_NAMES[i].second);
```
I think this works? not sure though

we also may as well restructure a bit and put this in a namespace function (the whole if(reload) thingy but I'm not sure if that's better or a lambda (or if the small replication is fine, mb)

Comment (RisingLeaf) [reply source/OutfitInfoDisplay.cpp:574]:
```suggestion
				attributeValues.emplace_back(Format::Number(60. * values[i] / reload) + VALUE_NAMES[i].second);
				if(fullDropoff != 1 && VALUE_NAMES[i].first.find("damage") != string::npos)
					attributesValues.emplace_back(" - " + Format::Number(60. * fullDropoff * values[i] / reload) + VALUE_NAMES[i].second);
```
consistency?

Comment (Quantumshark) [reply source/OutfitInfoDisplay.cpp]: Implemented in both places.

File: source/OutfitInfoDisplay.cpp [hunk 2]
```diff
@@ -561,6 +566,8 @@ void OutfitInfoDisplay::UpdateAttributes(const Outfit &outfit)
 			{
 				attributeLabels.emplace_back(VALUE_NAMES[i].first + PER_SECOND);
 				attributeValues.emplace_back(Format::Number(60. * values[i] / reload) + VALUE_NAMES[i].second);
+				if(fullDropoff != 1 && VALUE_NAMES[i].first.find("damage") != string::npos)
+					attributeValues.back() += " - " + Format::Number(60. * fullDropoff * values[i] / reload) + VALUE_NAMES[i].second;
```
Comment (RisingLeaf) [inline source/OutfitInfoDisplay.cpp:570]:
```suggestion
					attributeValues.emplace_back(" - " + Format::Number(60. * fullDropoff * values[i] / reload) + VALUE_NAMES[i].second);
```
this is what I meant with consistency, should have made myself clear

Comment (Quantumshark) [reply source/OutfitInfoDisplay.cpp:570]: That seems to break everything. 
image

Comment (Hurleveur) [reply source/OutfitInfoDisplay.cpp:570]: yeah you're adding another element if you do that @RisingLeaf which means a new line

Comment (RisingLeaf) [reply source/OutfitInfoDisplay.cpp:570]: ah, in that case dont do it

**Comment by Hurleveur** [comment]
> Doesn't currently work with submunitions, and given that currently that is (as far as I am aware) the only place damage dropoff is currently used in vanilla, I'd really like to actually get that working.

Yeah it is the only usage for it rn, not sure how to implement that tho so if anybody has ideas go ahead - mb add a check in the dropoff function that returns the submunition's dropoff if the main gun has no range? that could be a bit of a hack tho

**Comment by Hurleveur** [comment]
What about dropoff starting at say 500 range and ending at 1000 range?
We could improve how that is displayed later tbf

**Comment by Saturnalia9** [comment]
> `damage before dropoff - damage after dropoff`
> At point-blank range, your damage will always be the first number. At max range, it will be the second number.
> The exact shape of the graph in between depends on the exact stats, but it always consists of a single linear segment, which either covers the whole range or connects to flat lines at one or both ends.

So explain how this will be... explained? Damage after dropoff in reference to deceleration based on... subatomic resistances? Or due to potential weapon inaccuracy?

**Comment by Quantumshark** [comment]
> So explain how this will be... explained? Damage after dropoff in reference to deceleration based on... subatomic resistances? Or due to potential weapon inaccuracy?

To clarify, I'm not adding damage dropoff as a mechanic. That was done a long time ago. All this PR is doing is showing the player the damage dropoff on their weapon.

Weapons can give an explanation in their description if desired, but there's no need to try to come up with realistic explanations for the mechanics of damage dropoff. You can just say that the nature of a weapon makes it more effective at closer ranges.

And no, this isn't being used to model inaccuracy. We already have a way to model inaccuracy, it's called `inaccuracy`.

**Review by OcelotWalrus — APPROVED** [review]
It seems fair to display this kind of attribute.

**Comment by quyykk** [comment]
Hmm I'm not sure about this, mainly because of the lack of tooltip information and the fact that the line might be too long for certain weapons. It might be better to add additional lines for the dropoffs, that way we can also see the range. Thoughts?

**Comment by Amazinite** [comment]
How about adding a new attribute below "range" for dropoff weapons that describes when `range dropoff:` starts and ends. So taking the example from the PR description, this weapon would display `range dropoff: 0-510` below `range:`. The tooltip for `range dropoff:` would then describe how the hyphenated values mean for the damage attributes below.

Could you test what it looks like to have triple digit damage values with this for all damage types? If this causes the damage values to overlap with the attribute names, then we should probably have some different method of displaying the min and max. Perhaps we display damage types across two lines like this.
Attribute | Value
--|--
shield damage / second | 53
\- | 26.5
hull damage / second | 33
\- | 16.5

**Comment by Quantumshark** [comment]
image
Indeed, this does seem to cause overlapping (and would be much worse with disruption instead of slowing). I'll have a look at implementing your suggested approach.

**Comment by Quantumshark** [comment]
image

Bugs with my attempt to implement this notwithstanding, this does make the entry quite long. Is that an issue?

**Comment by quyykk** [comment]
Yeah way too long, and it's also a bit of an information overload. Honestly, I'd just display the values below range, maybe something like this:

image

**Comment by Quantumshark** [comment]
I don't actually know how to get the dropoff range in this section of the code. I think I've got dropoff modifier displaying properly, though.

**Comment by quyykk** [comment]
You'll need to add a new function to Weapon.

**Comment by Quantumshark** [comment]
image

File: data/_ui/tooltips.txt [hunk 3]
```diff
@@ -768,6 +768,12 @@ tip "ammo usage:"
 tip "range:"
 	`Total range of this weapon.`
 
+tip "dropoff modifier:"
+	`Multiplier to damage at max range.`
+
+tip "dropoff range:"
+	`Ranges between which the damage falls off from 100% to the displayed dropoff modifier.`
```
Comment (Amazinite) [inline data/_ui/tooltips.txt:775]: How do these sound?
```suggestion
tip "dropoff modifier:"
	`The multiplier applied to the damage of this weapon once its projectile has reached the maximum dropoff range.`

tip "dropoff range:"
	`The starting and ending range between which the damage of this weapon ramps from 100% to the displayed dropoff modifier.`
```

File: source/OutfitInfoDisplay.cpp [hunk 4]
```diff
@@ -471,6 +471,22 @@ void OutfitInfoDisplay::UpdateAttributes(const Outfit &outfit)
...
+		attributeValues.emplace_back(Format::Number(100 * fullDropoff) + "%");
+		attributesHeight += 20;
+		// Identify the ranges between which the dropoff takes place.
+		attributeLabels.emplace_back("dropoff range:");
+		attributeValues.emplace_back(Format::Number(outfit.DropoffStart())
+			+ " - " + Format::Number(outfit.DropoffEnd()));
+	}
+
+
+
```
Comment (Amazinite) [inline source/OutfitInfoDisplay.cpp:489]:
```suggestion
```

File: source/OutfitInfoDisplay.cpp [hunk 5]
```diff
@@ -471,6 +471,22 @@ void OutfitInfoDisplay::UpdateAttributes(const Outfit &outfit)
 	attributeValues.emplace_back(Format::Number(outfit.Range()));
 	attributesHeight += 20;
 
+	// Identify the dropoff at range and inform the player.
+	double fullDropoff = outfit.MaxDropoff();
+
```
Comment (Amazinite) [inline source/OutfitInfoDisplay.cpp:476]:
```suggestion
```

File: source/OutfitInfoDisplay.cpp [hunk 6]
```diff
@@ -471,6 +471,22 @@ void OutfitInfoDisplay::UpdateAttributes(const Outfit &outfit)
...
+	if(fullDropoff != 1.)
+	{
+		attributeLabels.emplace_back("dropoff modifier:");
+		attributeValues.emplace_back(Format::Number(100 * fullDropoff) + "%");
```
Comment (Amazinite) [inline source/OutfitInfoDisplay.cpp:480]:
```suggestion
		attributeValues.emplace_back(Format::Number(100. * fullDropoff) + "%");
```

File: source/Weapon.cpp [hunk 7]
```diff
@@ -511,6 +511,29 @@ double Weapon::DamageDropoff(double distance) const
...
+{
+	return damageDropoffRange.first;
+}
+
+
+
+// Return the range at which the weapon's damage dropoff ends
+double Weapon::DropoffEnd() const
+{
+	return (damageDropoffRange.second ? damageDropoffRange.second : Range());
```
Comment (Amazinite) [inline source/Weapon.cpp:533]: A sanity check in the load function already sets the second value to the range of the weapon.
```suggestion
	return damageDropoffRange.second;
```

File: source/OutfitInfoDisplay.cpp [hunk 8]
```diff
@@ -471,6 +471,22 @@ void OutfitInfoDisplay::UpdateAttributes(const Outfit &outfit)
...
+		attributesHeight += 20;
+		// Identify the ranges between which the dropoff takes place.
+		attributeLabels.emplace_back("dropoff range:");
+		attributeValues.emplace_back(Format::Number(outfit.DropoffStart())
+			+ " - " + Format::Number(outfit.DropoffEnd()));
```
Comment (Amazinite) [inline source/OutfitInfoDisplay.cpp:485]: Are we missing an `attributesHeight += 20;` here?

Comment (Quantumshark) [reply source/OutfitInfoDisplay.cpp:484]: Good catch.

File: source/Weapon.h [hunk 9]
```diff
@@ -192,6 +192,12 @@ class Weapon {
 	// Calculate the percent damage that this weapon deals given the distance
 	// that the projectile traveled if it has a damage dropoff range.
 	double DamageDropoff(double distance) const;
+	// Return the weapon's damage dropoff at maximum range
+	double MaxDropoff() const;
+	// Return the range at which the weapon's damage dropoff begins
+	double DropoffStart() const;
+	// Return the range at which the weapon's damage dropoff ends
+	double DropoffEnd() const;
```
Comment (quyykk) [inline source/Weapon.h:200]: No need to have separate functions for the start and the end when the variable is stored as a range/pair internally anyways.

Comment (Amazinite) [inline source/Weapon.h:200]:
```suggestion
	// Return the ranges at which the weapon's damage dropoff begins and ends.
	const std::pair<double, double> &DropoffRanges() const;
```

File: source/Weapon.cpp [hunk 10]
```diff
@@ -511,6 +511,29 @@ double Weapon::DamageDropoff(double distance) const
...
+	return damageDropoffRange.first;
+}
+
+
+
+// Return the range at which the weapon's damage dropoff ends
+double Weapon::DropoffEnd() const
+{
+	return damageDropoffRange.second;
+}
```
Comment (Amazinite) [inline source/Weapon.cpp:534]:
```suggestion
// Return the ranges at which the weapon's damage dropoff begins and ends.
pair<double, double> &Weapon::DropoffRanges() const
{
	return damageDropoffRange;
}
```

File: source/OutfitInfoDisplay.cpp [hunk 11]
```diff
@@ -474,6 +474,20 @@ void OutfitInfoDisplay::UpdateAttributes(const Outfit &outfit)
...
+	double fullDropoff = outfit.MaxDropoff();
+	if(fullDropoff != 1.)
+	{
+		attributeLabels.emplace_back("dropoff modifier:");
+		attributeValues.emplace_back(Format::Number(100. * fullDropoff) + "%");
+		attributesHeight += 20;
+		// Identify the ranges between which the dropoff takes place.
+		attributeLabels.emplace_back("dropoff range:");
+		attributeValues.emplace_back(Format::Number(outfit.DropoffStart())
+			+ " - " + Format::Number(outfit.DropoffEnd()));
```
Comment (Amazinite) [inline source/OutfitInfoDisplay.cpp:487]:
```suggestion
		const pair<double, double> &ranges = outfit.DropoffRanges();
		attributeValues.emplace_back(Format::Number(ranges.first)
			+ " - " + Format::Number(ranges.second));
```

File: source/Weapon.cpp [hunk 12]
```diff
@@ -511,6 +511,21 @@ double Weapon::DamageDropoff(double distance) const
...
+// Return the weapon's damage dropoff at maximum range
+double Weapon::MaxDropoff() const
+{
+	return damageDropoffModifier;
+}
+
+
+
+// Return the ranges at which the weapon's damage dropoff begins and ends.
+pair<double, double> &Weapon::DropoffRanges() const
```
Comment (Amazinite) [inline source/Weapon.cpp:523]: Oops. 😳 
```suggestion
const pair<double, double> &Weapon::DropoffRanges() const
```

File: source/Weapon.h [hunk 13]
```diff
@@ -192,6 +192,10 @@ class Weapon {
 	// Calculate the percent damage that this weapon deals given the distance
 	// that the projectile traveled if it has a damage dropoff range.
 	double DamageDropoff(double distance) const;
+	// Return the weapon's damage dropoff at maximum range
```
Comment (Amazinite) [inline source/Weapon.h:195]:
```suggestion
	// Return the weapon's damage dropoff at maximum range.
```

File: source/Weapon.cpp [hunk 14]
```diff
@@ -511,6 +511,21 @@ double Weapon::DamageDropoff(double distance) const
 }
 
 
+// Return the weapon's damage dropoff at maximum range
```
Comment (Amazinite) [inline source/Weapon.cpp:514]:
```suggestion

// Return the weapon's damage dropoff at maximum range.
```

**Pull request merged**